ntoskrnl/mm/i386/procsup.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * PROJECT:         ReactOS Kernel
+ * LICENSE:         BSD-3-Clause (https://spdx.org/licenses/BSD-3-Clause.html)
+ * FILE:            ntoskrnl/mm/i386/procsup.c
+ * PURPOSE:         Process handling for i386 architecture
+ * PROGRAMMERS:     Jérôme Gardou
+ */
+
+/* INCLUDES *******************************************************************/
+
+#include <ntoskrnl.h>
+#define NDEBUG
+#include <debug.h>
+
+#define MODULE_INVOLVED_IN_ARM3
+#include <mm/ARM3/miarm.h>
+
+BOOLEAN
+MiArchCreateProcessAddressSpace(
+    _In_ PEPROCESS Process,
+    _In_ PULONG_PTR DirectoryTableBase)
+{
+    PFN_NUMBER PdeIndex = DirectoryTableBase[0] >> PAGE_SHIFT;
+    PFN_NUMBER HyperIndex = DirectoryTableBase[1] >> PAGE_SHIFT;
+    PMMPTE PointerPte;
+    MMPTE PdePte, TempPte;
+    PMMPTE PteTable;
+    ULONG PdeOffset;
+    KIRQL OldIrql;
+
+    /* Get a PTE  */
+    PointerPte = MiReserveSystemPtes(1, SystemPteSpace);
+    if (!PointerPte)
+        return FALSE;
+    PteTable = MiPteToAddress(PointerPte);
+
+    /* Build a page table for hyper space */
+    MI_MAKE_HARDWARE_PTE_KERNEL(&PdePte,
+                                PointerPte,
+                                MM_READWRITE,
+                                HyperIndex);
+
+    /* Set it dirty and map it */
+    MI_MAKE_DIRTY_PAGE(&PdePte);
+    MI_WRITE_VALID_PTE(PointerPte, PdePte);
+
+
+    /* Now write the PTE/PDE entry for the working set list index itself */
+    TempPte = ValidKernelPteLocal;
+    TempPte.u.Hard.PageFrameNumber = Process->WorkingSetPage;
+    PdeOffset = MiAddressToPteOffset(MmWorkingSetList);
+    PteTable[PdeOffset] = TempPte;
+
```

Comment:
Is it correct to assume that here you did some sort of optimization by not calling `MiReleaseSystemPtes()` and then later re-calling `MiReserveSystemPtes()`, as was the case in the old code (in `MmCreateProcessAddressSpace()`)?

Reply:
Yes. "Some sort of optimization" if you like to call it that way. I'd prefer to say "not doing something obviously suboptimal".